server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -86,19 +85,13 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
             return YES_AUTO_EXPAND_ALL;
         }
         if (node.node() != null) {
-            assert Strings.hasLength(node.node().getHostAddress()) : node;
-            for (RoutingNode checkNode : allocation.routingNodes()) {
-                if (checkNode.node() == null) {
-                    continue;
-                }
-                // check if its on the same host as the one we want to allocate to
-                assert Strings.hasLength(checkNode.node().getHostAddress()) : checkNode;
```

Comment:
Similarly for this assert, you will probably have to extract 

```
allocation.nodes()
                    .getDataNodes()
                    .get(assignedShard.currentNodeId())
                    .getHostAddress()
```

below into a local variable to be able to also assert on it.